stdlib/result.mli

```diff
@@ -0,0 +1,96 @@
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Result values.
+
+    Result values handle computation results and errors in an explicit
+    and declarative manner without resorting to exceptions.
+
+    @since 4.08 *)
+
+(** {1:results Results} *)
+
+type ('a, 'e) t = ('a, 'e) result = Ok of 'a | Error of 'e
+(** The type for result values. Either a value [Ok v] or an error [Error e]. *)
+
+val ok : 'a -> ('a, 'e) result
+(** [ok v] is [Ok v]. *)
+
+val error : 'e -> ('a, 'e) result
+(** [error e] is [Error e]. *)
+
+val value : ('a, 'e) result -> default:'a -> 'a
+(** [value r ~default] is [v] if [r] is [Ok v] and [default] otherwise. *)
+
+val get_ok : ('a, 'e) result -> 'a
+(** [get_ok r] is [v] if [r] is [Ok v] and @raise Invalid_argument
+    otherwise. *)
+
+val get_error : ('a, 'e) result -> 'e
+(** [get_error r] is [e] if [r] is [Error e] and @raise Invalid_argument
+    otherwise. *)
+
+val bind : ('a, 'e) result -> ('a -> ('b, 'e) result) -> ('b, 'e) result
+(** [bind r f] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val join : (('a, 'e) result, 'e) result -> ('a, 'e) result
+(** [join rr] is [r] if [rr] is [Ok r] and [rr] if [rr] is [Error _]. *)
+
+val map : ('a -> 'b) -> ('a, 'e) result -> ('b, 'e) result
+(** [map f r] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val fold : ('a -> 'c) -> ('e -> 'c) -> ('a, 'e) result -> 'c
+(** [fold ok error r] is [ok v] if [r] is [Ok v] and [error e] if [r]
+    is [Error e]. *)
+
+val iter : ('a -> unit) -> ('a, 'e) result -> unit
+(** [iter f r] is [f v] if [r] is [Ok v] and [()] otherwise. *)
+
+val recover : ('e -> 'a) -> ('a, 'e) result -> ('a, 'f) result
+(** [recover f r] is [r] if [r] is [Ok v] and [Ok (f e)] if [r] is
+    [Error e]. *)
+
+(** {1:preds Predicates and comparisons} *)
+
+val is_ok : ('a, 'e) result -> bool
+(** [is_ok r] is [true] iff [r] is [Ok _]. *)
+
+val is_error : ('a, 'e) result -> bool
+(** [is_error r] is [true] iff [r] is [Error _]. *)
+
+val equal :
+  ok:('a -> 'a -> bool) -> error:('e -> 'e -> bool) -> ('a, 'e) result ->
+  ('a, 'e) result -> bool
+(** [equal ~ok ~error r0 r1] tests equality of [r0] and [r1] using [ok]
+    and [error] to respectively compare values wrapped by [Ok _] and
+    [Error _]. *)
+
+val compare :
+  ok:('a -> 'a -> int) -> error:('e -> 'e -> int) -> ('a, 'e) result ->
+  ('a, 'e) result -> int
+(** [compare ~ok ~error r0 r1] totally orders [r0] and [r1] using [ok] and
+    [error] to respectively compare values wrapped by [Ok _ ] and [Error _].
+    [Ok _] values are smaller than [Error _] values. *)
```

Comment:
Let's be coherent with `Option`, and put `Error` before `Ok` (or change the PR about `Option`).

Reply:
I prefer to follow the order of the cases in the type definitions.

Reply:
Ah sorry I misunderstood that one. I thought you were talking about argument order of the signature. My comment might still apply even though I'm less assertive about it.

Reply:
On the plus side of the current def. is that assuming [ok] and [error] are compatible with polymorphic compare, the resulting comparison will be. This might be "useful" if you mistakingly use polymorphic comparison instead of the specialized comparison you should have used.